runtime/interpreter/encode.go

```diff
@@ -562,22 +617,85 @@ type encodedDictionaryValue struct {
 	Entries map[string]interface{} `cbor:"1,keyasint"`
 }
 
+const dictionaryKeyPathPrefix = "k"
+const dictionaryValuePathPrefix = "v"
+
 // TODO: optimize: use CBOR map, but unclear how to preserve ordering
-func (e *Encoder) prepareDictionaryValue(v *DictionaryValue) (interface{}, error) {
-	keys, err := e.prepareArray(v.Keys)
+func (e *Encoder) prepareDictionaryValue(
+	v *DictionaryValue,
+	path []string,
+	deferrals *EncodingDeferrals,
+) (
+	interface{},
+	error,
+) {
+	keysPath := append(path[:], dictionaryKeyPathPrefix)
+
+	keys, err := e.prepareArray(v.Keys, keysPath, deferrals)
 	if err != nil {
 		return nil, err
 	}
 
 	entries := make(map[string]interface{}, len(v.Entries))
 
+	// Deferring the encoding of values is only supported if all
+	// values are resources: resource typed dictionaries are moved
+
+	deferred := e.deferred
+	if deferred {
+		for _, value := range v.Entries {
+			compositeValue, ok := value.(*CompositeValue)
+			if !ok || compositeValue.Kind != common.CompositeKindResource {
+				deferred = false
+				break
+			}
+		}
+	}
+
 	for _, keyValue := range v.Keys.Values {
 		key := dictionaryKey(keyValue)
-		prepared, err := e.prepare(v.Entries[key])
-		if err != nil {
-			return nil, err
+		entryValue := v.Entries[key]
+		valuePath := append(path[:], dictionaryValuePathPrefix, key)
+
+		if deferred {
+
+			deferredStorageKey, isDeferred := v.DeferredKeys[key]
+
+			// If the value is not deferred, i.e. it is in memory,
+			// then it must be stored under a separate storage key
+			// in the owner's storage.
+
+			if !isDeferred {
+				deferrals.Values[joinPath(valuePath)] = entryValue
+			} else {
+
+				// If the value is deferred, and the deferred value
+				// is stored in another account's storage,
+				// it must be moved.
+
+				deferredOwner := *v.DeferredOwner
+				owner := *v.Owner
+
+				if deferredOwner != owner {
+
+					deferrals.Moves = append(deferrals.Moves,
```

Comment:
Would be nice to see a couple unit tests around the deferred moves if possible! Specifically around `DeferredOwner` / `NewOwner`.